isobus/include/isobus/isobus/can_parameter_group_builder.hpp

```diff
@@ -0,0 +1,503 @@
+			*data = 0;
+			while (bits)
+			{
+				// Move in the new data.
+				*data = (*data & (255 >> space)) | ((buffer[byte] >> input) << output);
+				if (space >= bits)
+				{
+					space -= bits;
+					break;
+				}
+				if (space > remaining)
+				{
+					space -= remaining;
+					output = 8 - space;
+					remaining = 0;
+					input = 0;
+					++byte;
+				}
+				else if (remaining > space)
+				{
+					bits -= space;
+					remaining -= space;
+					input = 8 - remaining;
+					space = 8;
+					output = 0;
+					++data;
+					*data = 0;
+				}
+				else
+				{
+					bits -= space;
+					// Perfectly matched.
+					remaining = 8;
+					input = 0;
+					space = 8;
+					output = 0;
+					++data;
+					*data = 0;
+					++byte;
+				}
+			}
+			if (space)
+			{
+				// Bits left to mask.
+				*data = *data & (255 >> space);
+			}
+			return true;
+		}
+
+	public:
+		ParameterGroupBuilder() :
+		  buffer()
+		{
+		}
+
+		ParameterGroupBuilder(std::vector<std::uint8_t> const &data) :
+		  buffer()
+		{
+			buffer = data;
+			writeOffset = data.size() * 8;
+		}
+
+		std::size_t get_written_bits() const
+		{
+			return writeOffset;
+		}
+
+		std::size_t get_written_bytes() const
+		{
+			return (writeOffset + 7) / 8;
+		}
+
+		std::size_t get_read_bits() const
+		{
+			return readOffset;
+		}
+
+		std::size_t get_read_bytes() const
+		{
+			return (readOffset + 7) / 8;
+		}
+
+		template <typename T>
+		bool write(T const & data)
+		{
+			return write_bits((std::uint8_t const *)&data, sizeof (T) * 8);
+		}
+
+		template <typename T>
+		bool write(T const & data, std::size_t bits)
+		{
+			return write_bits((std::uint8_t const *)&data, bits);
+		}
+
+		template <>
+		bool write<bool>(bool const & data)
+		{
+			std::uint8_t bits = data ? 255 : 0;
+			return write_bits(&bits, 1);
+		}
+
+		template <>
+		bool write<char const *>(char const * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		template <>
+		bool write<char *>(char * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		template <>
+		bool write<std::uint8_t const *>(std::uint8_t const * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		template <>
+		bool write<std::uint8_t *>(std::uint8_t * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		bool write(char const * data, bool includeNull)
+		{
+			return write((std::uint8_t const *)data, includeNull);
+		}
+
+		bool write(char * data, bool includeNull)
+		{
+			return write((std::uint8_t const *)data, includeNull);
+		}
+
+		bool write(std::uint8_t const * data, bool includeNull)
+		{
+			// Base case.  Write each byte separately so they don't get put in little-endian, which
```

Comment:
Little-endian is the normal way in the standard, so probably meant `big` in this comment